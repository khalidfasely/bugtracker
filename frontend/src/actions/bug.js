import bugApi from "../api/bug";

export const setBug = (bug) => ({
    type: "SET_BUG",
    bug
});

export const startSetBug = (pid, bid) => {
    return (dispatch) => {
        return bugApi(pid, bid).then((result) => {
            dispatch(setBug(result.bug));
            return result;
        });
    };
};