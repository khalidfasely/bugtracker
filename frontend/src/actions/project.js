import delBugApi from "../api/delBug";
import newBugApi from "../api/newBug";
import projectApi from "../api/project";

export const setProject = (project, bugs) => ({
    type: "SET_PROJECT",
    project,
    bugs
});

export const startSetProject = (pid) => {
    return (dispatch) => {
        return projectApi(pid).then((result) => {
            dispatch(setProject(result.project, result.bugs));
            return result;
        });
    };
};

export const setNewBug = (bug) => ({
    type: 'NEW_BUG',
    bug
});

export const startSetNewBug = (on_project, bugData) => {
    return (dispatch) => {
        return newBugApi(on_project, bugData).then(result => {
            if (result.bug) {
                dispatch(setNewBug(result.bug));
            }
            return result;
        });
    };
};

export const setDeleteBug = (bid) => ({
    type: 'DELETE_BUG',
    bid
});

export const startSetDeleteBug = (bid) => {
    return (dispatch) => {
        return delBugApi(bid).then(result => {
            dispatch(setDeleteBug(bid));
            return result;
        });
    };
};