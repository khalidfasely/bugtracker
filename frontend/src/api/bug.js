export default async (pid, bid) => {
    const res = await fetch(`/api/project/${pid}/bug/${bid}`);

    const result = await res.json();

    return result;
};