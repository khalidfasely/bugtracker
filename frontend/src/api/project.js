export default async (pid) => {
    const res = await fetch(`/api/project/${pid}`);

    const project = await res.json();

    return project;
};