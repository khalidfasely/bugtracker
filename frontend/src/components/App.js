function App() {
  return (
    <div className="App">
      <header>
        Hello world from React        
      </header>
    </div>
  );
}

export default App;
