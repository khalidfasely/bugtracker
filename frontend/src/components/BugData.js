const BugData = ({ bug }) => (
    <div>
        <p>{bug?.id}</p>
        <p>{bug?.user?.username}</p>
        <p>{bug?.title}</p>
        <p>{bug?.description}</p>
        <p>{bug?.time}</p>
        <p>{bug?.classification}</p>
        <p>{`${bug?.active}`}</p>
        <div>
            {bug?.admins?.map(admin => <span key={admin} >{admin} </span>)}
        </div>
        <br />
        <div>
            {bug?.users_with?.map(user => <span key={user} >{user} </span>)}
        </div>
    </div>
);

export default BugData;