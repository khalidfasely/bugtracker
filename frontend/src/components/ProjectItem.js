import { Link } from "react-router-dom";
import { connect } from "react-redux";
import ProjectData from "./ProjectData";
import BugsList from "./BugsList";
import NewBug from "./NewBug";

const ProjectItem = ({uname, projectItem, bugs}) => {
    return (
        <>
            {
                projectItem?.users_with?.includes(uname) ?
                <div>
                    **** project <br /><ProjectData projectItem={projectItem} />
                    **** new bug <br /><NewBug users={projectItem.users_with} on_project={projectItem.id} />
                    **** bugs <br /><div>
                        { bugs?.map(bug => (
                            <Link to={`/project/${bug?.on_project}/bug/${bug?.id}`}>
                                <BugsList bug={bug} />
                            </Link>)
                        )}
                    </div>
                </div> :
                <div>Not able to see this project! <Link to="/new">Create one!</Link></div>
            }
        </>
    );
}

const mapStateToProps = (state) => ({
    uname: state.auth.uname,
    projectItem: state.projectItem.project,
    bugs: state.projectItem.bugs
});

export default connect(mapStateToProps)(ProjectItem);