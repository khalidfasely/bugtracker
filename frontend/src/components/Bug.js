import { useState, useEffect } from "react";
import { connect } from "react-redux";
import { useParams } from "react-router-dom";
import { startSetBug } from "../actions/bug";

import BugItem from './BugItem';

export const Bug = ({ startSetBug }) => {
    const params = useParams();

    const [ableToSee, setAbleToSee] = useState(false);
    const [notFound, setNotFound] = useState('');

    useEffect(() => {
        startSetBug(params.pid, params.bid)
        .then((result) => {
            setNotFound('');

            if (result.message) {
                setNotFound(result.message);
            };
        })
        .then(() => setAbleToSee(true))
        .catch(er => console.error(er));
    }, [params])


    if (notFound) {
        return <div>Bug Not Found! -- {notFound}</div>;
    };

    return (
        <div>
            {
                ableToSee ?
                <div><BugItem /></div> :
                <div>You're not able to see this bug</div>
            }
        </div>
    );
};

const mapDispatchToProps = (dispatch) => ({
    startSetBug: (pid, bid) => dispatch(startSetBug(pid, bid))
});

export default connect(undefined, mapDispatchToProps)(Bug);