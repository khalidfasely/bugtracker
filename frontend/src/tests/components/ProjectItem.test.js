import React from "react";
import { render } from "@testing-library/react";

import { ProjectItem } from "../../components/ProjectItem";
import project from '../fixtures/project';
import bug from '../fixtures/bug';

jest.mock('../../components/ProjectDelModal', () => () => <div>ProjectDelModal</div>);
jest.mock('../../components/ProjectEditModal', () => () => <div>ProjectEditModal</div>);
jest.mock('../../components/NewBug', () => () => <div>NewBug</div>);
//jest.mock('../../components/BugsList', () => () => <div>BugsList</div>);

let props;

beforeEach(() => {//uname, projectItem, bugs, startSetDelProject
    props = {
        uname: 'User',
        projectItem: project,
        bugs: [bug],
        startSetDelProject: jest.fn(() => Promise.resolve())
    }
});

test('Should render ProjectItem component correctly with uname not in the project', () => {
    const { asFragment } = render(<ProjectItem {...props} />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render ProjectItem component correctly with uname in the project', () => {
    const { asFragment } = render(<ProjectItem {...{...props, uname: 'admin'}} />);
    expect(asFragment()).toMatchSnapshot();
});

test('Should render ProjectItem component correctly with uname === project.user', () => {
    const { asFragment } = render(<ProjectItem {...{...props, uname: project.user.username}} />);
    expect(asFragment()).toMatchSnapshot();
});