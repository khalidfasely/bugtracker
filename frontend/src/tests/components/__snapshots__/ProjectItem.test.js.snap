// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render ProjectItem component correctly with uname === project.user 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div>
        <p>
          1
        </p>
        <p>
          Project one
        </p>
        <p>
          now
        </p>
        <p>
          1
        </p>
        <div>
          <span>
            1 
          </span>
        </div>
        <br />
        <div>
          <span>
            1 
          </span>
          <span>
            admin 
          </span>
        </div>
      </div>
      <hr />
      <div>
        <button>
          Edit
        </button>
        <button>
          Delete
        </button>
      </div>
      <hr />
      <div>
        NewBug
      </div>
      <hr />
      <div>
        <a>
          <div>
            <p>
              1
            </p>
            <p>
              Bug title
            </p>
            <p>
              medium
            </p>
            <p>
              now
            </p>
            <p>
              admin
            </p>
            <div>
              <span>
                admin 
              </span>
              <span>
                1 
              </span>
            </div>
            <br />
            <div>
              <span>
                admin 
              </span>
              <span>
                1 
              </span>
            </div>
          </div>
        </a>
      </div>
    </div>
    <div>
      ProjectDelModal
    </div>
    <div>
      ProjectEditModal
    </div>
  </div>
</DocumentFragment>
`;

exports[`Should render ProjectItem component correctly with uname in the project 1`] = `
<DocumentFragment>
  <div>
    <div>
      <div>
        <p>
          1
        </p>
        <p>
          Project one
        </p>
        <p>
          now
        </p>
        <p>
          1
        </p>
        <div>
          <span>
            1 
          </span>
        </div>
        <br />
        <div>
          <span>
            1 
          </span>
          <span>
            admin 
          </span>
        </div>
      </div>
      <hr />
      <hr />
      <div>
        NewBug
      </div>
      <hr />
      <div>
        <a>
          <div>
            <p>
              1
            </p>
            <p>
              Bug title
            </p>
            <p>
              medium
            </p>
            <p>
              now
            </p>
            <p>
              admin
            </p>
            <div>
              <span>
                admin 
              </span>
              <span>
                1 
              </span>
            </div>
            <br />
            <div>
              <span>
                admin 
              </span>
              <span>
                1 
              </span>
            </div>
          </div>
        </a>
      </div>
    </div>
    <div>
      ProjectDelModal
    </div>
    <div>
      ProjectEditModal
    </div>
  </div>
</DocumentFragment>
`;

exports[`Should render ProjectItem component correctly with uname not in the project 1`] = `
<DocumentFragment>
  <div>
    <div>
      Not able to see this project! 
      <a>
        Create one!
      </a>
    </div>
    <div>
      ProjectDelModal
    </div>
    <div>
      ProjectEditModal
    </div>
  </div>
</DocumentFragment>
`;
